lib/VideoInputTest.ts

```diff
@@ -0,0 +1,372 @@
+} from './types';
+import {
+  InvalidityRecord,
+  validateDeviceId,
+  validateOptions,
+  validateTime,
+} from './utils/optionValidation';
+
+export declare interface VideoInputTest {
+  /**
+   * This event is emitted with a test report when the test ends.
+   * @param event [[VideoInputTest.Events.End]]
+   * @param report Summary of the test.
+   * @private
+   */
+  emit(
+    event: VideoInputTest.Events.End,
+    report: VideoInputTest.Report,
+  ): boolean;
+  /**
+   * This event is emitted with a [[DiagnosticError]] when the test encounters
+   * an error, fatal or not.
+   * @param event [[VideoInputTest.Events.Error]]
+   * @param error The [[DiagnosticError]] that was encountered.
+   * @private
+   */
+  emit(
+    event: VideoInputTest.Events.Error,
+    error: DiagnosticError,
+  ): boolean;
+
+  /**
+   * Raised upon completion of the test.
+   * @param event [[VideoInputTest.Events.End]]
+   * @param listener A callback that expects the following parameters:
+   *  An [[VideoInputTest.Report]] that summarizes the test.
+   * @returns This [[VideoInputTest]] instance.
+   * @event
+   */
+  on(
+    event: VideoInputTest.Events.End,
+    listener: (report: VideoInputTest.Report) => any,
+  ): this;
+  /**
+   * Raised by the test when encountering an error with a parameter of type
+   * [[DiagnosticError]].
+   * @param event [[VideoInputTest.Events.Error]]
+   * @param listener A callback that expects the following parameters:
+   *  A [[DiagnosticError]] that the test encountered.
+   * @returns This [[VideoInputTest]] instance.
+   * @event
+   */
+  on(
+    event: VideoInputTest.Events.Error,
+    listener: (error: DiagnosticError) => any,
+  ): this;
+}
+
+/**
+ * [[VideoInputTest]] class that parses options and starts a video input device
+ * test.
+ *
+ * Please see [[testVideoInputDevice]] for details and recommended practices.
+ */
+export class VideoInputTest extends EventEmitter {
+  /**
+   * The name of the test.
+   */
+  static testName: string = 'video-input-test';
+  /**
+   * Default options for the test.
+   */
+  private static defaultOptions: VideoInputTest.InternalOptions = {
+    debug: false,
+    duration: Infinity,
+    getUserMedia,
+  };
+
+  /**
+   * Teststamp of when the test was ended.
+   */
+  private _endTime: number | null = null;
+  /**
+   * An array of any errors that occur during the run time of the test.
+   */
+  private readonly _errors: DiagnosticError[] = [];
+  /**
+   * Options to be used throughout the runtime of the test.
+   */
+  private _options: VideoInputTest.InternalOptions;
+  /**
+   * The promise returned by the `HTMLMediaElement` after playing the stream
+   * captured by `getUserMedia`.
+   */
+  private _playPromise: Promise<void> | null = null;
+  /**
+   * Teststamp of when the test was constructed and started.
+   */
+  private _startTime: number;
+  /**
+   * Timer ID of the test duration timeout.
+   */
+  private _timeoutId: NodeJS.Timeout | null = null;
```

Comment:
curious - why `NodeJS.Timeout` ? is there some browser type that we can use ?

Reply:
There were some issues using `number` for previous instances of saving `setTimeout` return values so we resorted to `NodeJS.Timeout`.